std/textproto/writer.ts

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+// Based on https://github.com/golang/go/blob/92c732e901a732855f4b813e6676264421eceae9/src/net/textproto/writer.go
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+import { encode } from "../encoding/utf8.ts";
+import { BufWriter } from "../io/bufio.ts";
+
+type WriteState = 0 | 1 | 2 | 3;
+const WRITE_STATE_BEGIN = 0;
+const WRITE_STATE_BEGIN_LINE = 1;
+const WRITE_STATE_CR = 2;
+const WRITE_STATE_DATA = 3;
+
+const DOT = encode(".");
+const CR = encode("\r");
+const LF = encode("\n");
+const DOTCRLF = encode(".\r\n");
+
+export interface AsyncCloser {
+  close(): Promise<void>;
+}
+
+class DotWriter implements Deno.Writer, AsyncCloser {
+  #state: WriteState = WRITE_STATE_BEGIN;
+
+  constructor(readonly w: TextProtoWriter) {}
+
+  async write(p: Uint8Array): Promise<number> {
+    const bufWriter = this.w.bufWriter;
+    let n = 0;
+    while (n < p.length) {
+      const c = p[n];
+      switch (this.#state) {
+        case WRITE_STATE_BEGIN:
+        case WRITE_STATE_BEGIN_LINE:
+          this.#state = WRITE_STATE_DATA;
+          if (c === charCode(".")) {
+            await bufWriter.write(DOT);
+          }
+          // FALLTHROUGH
+        case WRITE_STATE_DATA:
+          if (c === charCode("\r")) {
+            this.#state = WRITE_STATE_CR;
+          }
+          if (c === charCode("\n")) {
+            await bufWriter.write(CR);
+            this.#state = WRITE_STATE_BEGIN_LINE;
+          }
+          break;
+        case WRITE_STATE_CR:
+          this.#state = WRITE_STATE_DATA;
+          if (c === charCode("\n")) {
```

Comment:
More uses of the \n and \r.  As these are also reference in the other files can they share a enum to improve readability